### DISCUSSION TITLE
TableView - RowFilter

### DISCUSSION THREAD

**Discussion by aaronhudon** [tui-cs/Terminal.Gui#2377] [Q&A] [answered]
Does the TableView support the DataTable's underlying RowFilter?

**Comment by tznind** [comment, marked as answer]
`RowFilter` is part of `DefaultView` which is not used by `TableView`.  However you can generate a new table from the view (see code below for example).  If you have a very large table though you may not want to do this on every keystroke.  If you apply this filter using background threads or async make sure to update the table using `Application.MainLoop.Invoke` to prevent changing the table during drawing.  Also when you activate or right click etc the cells in the filtered table then the indexes will refer to what is currently displayed (not your original table e.g. `dt`).

Also **you will want some try/catch logic when setting** `RowFilter` because it gives exceptions on bad syntax which would go badly for you if unhandled in a `TextChanged` callback.

![tableview-filter](https://user-images.githubusercontent.com/31306100/220624260-111eee53-3430-4bd7-bf1a-cc4a63159133.gif)


```csharp
var find = new TextField{Width = 10};
find.TextChanged += (_)=>
{
    if(find.Text.Length == 0)
    {
        dt.DefaultView.RowFilter = null;
    }
    else
    {    
        dt.DefaultView.RowFilter = "There like '%" + find.Text.ToString() +"%'";
    }
    tv.Table = dt.DefaultView.ToTable();
    tv.Update();
};

win.Add(find);
```

**Reply by aaronhudon** [reply]
Thanks for the reply - too bad it does not support the DefaultView - my use case is thousands of rows.

**Reply by tznind** [reply]
Sorry, I should clarify my point on performance.   

`TableView` is designed to support millions of rows and hundreds of columns so may still be a good option for you.  Basically if `System.Data.DataTable` can handle it then `TableView` can.  For example rendering is done starting from the table cell scroll offsets (row/col) and fills the available console screen space (so its super fast and not affected by data volume).  

Any performance issues with the above code would be in the allocations of `System.Data.DataTable` and the `ToTable` method.

From an architecture point of view, you may want to apply filtering when you query your data source rather than into the UI element.  e.g. when user changes filter then send a new query to data source and update table with the results.  This moves processing to the server, reduces the amount of records held locally and typically would give you way more control over the filtering than `RowFilter` alone.
